tests/schema/tests.py

```diff
@@ -2481,6 +2488,287 @@ def test_text_field_with_db_index(self):
         assertion = self.assertIn if connection.features.supports_index_on_text_field else self.assertNotIn
         assertion('text_field', self.get_indexes(AuthorTextFieldWithIndex._meta.db_table))
 
+    @skipUnlessDBFeature('supports_expression_indexes')
+    def test_index_with_f_expressions(self):
+        with connection.schema_editor() as editor:
+            editor.create_model(Tag)
+        index = Index('slug', F('title').desc(), name='field_exp_idx')
+        # Add index.
+        with connection.schema_editor() as editor:
+            editor.add_index(Tag, index)
+            sql = index.create_sql(Tag, editor)
+        table = Tag._meta.db_table
+        self.assertIn(index.name, self.get_constraints(table))
+        if connection.features.supports_index_column_ordering:
+            self.assertIndexOrder(Tag._meta.db_table, index.name, ['ASC', 'DESC'])
+        # SQL contains columns.
+        self.assertIs(sql.references_column(table, 'slug'), True)
+        self.assertIs(sql.references_column(table, 'title'), True)
+        # Remove index.
+        with connection.schema_editor() as editor:
+            editor.remove_index(Tag, index)
+        self.assertNotIn(index.name, self.get_constraints(table))
+
+    @skipUnlessDBFeature('supports_expression_indexes', 'supports_index_collations')
```

Comment:
@hannseman  I don't think we need the `supports_index_collations` feature flag. MySQL supports collations in indexes, see an example from the MySQL [documentation](https://dev.mysql.com/doc/refman/8.0/en/create-index.html#create-index-functional-key-parts): 
```sql
INDEX idx ((CAST(data->>"$.name" AS CHAR(30)) COLLATE utf8mb4_bin))
```
Wrapping in parentheses is missing for `Collate` and as a consequence we generate an invalid SQL:
```sql
CREATE INDEX `composite_collate_idx` ON `schema_book` (
    `title` COLLATE `utf8_esperanto_ci` DESC,
    `slug` COLLATE `utf8_esperanto_ci`
)
```
Wrapped collations works fine on MySQL:
```sql
CREATE INDEX `composite_collate_idx` ON `schema_book` (
    (`title` COLLATE `utf8_esperanto_ci`) DESC,
    (`slug` COLLATE `utf8_esperanto_ci`)
) 
```

Reply:
We could always wrap collations in parentheses :thinking: 
```diff
diff --git a/django/db/models/functions/comparison.py b/django/db/models/functions/comparison.py
index 8a1c34430b..8b4d499c89 100644
--- a/django/db/models/functions/comparison.py
+++ b/django/db/models/functions/comparison.py
@@ -79,7 +79,7 @@ class Coalesce(Func):
 
 class Collate(Func):
     function = 'COLLATE'
-    template = '%(expressions)s %(function)s %(collation)s'
+    template = '(%(expressions)s %(function)s %(collation)s)'
     # Inspired from https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
     collation_re = _lazy_re_compile(r'^[\w\-]+$')
 

```

Reply:
Ah wow, good catch. I tried to find information about MySQL support on `CREATE INDEX` but found none. Looks like MySQL treats the usage of `COLLATE` as a regular functional index, while SQLite and PostgreSQL treats it as a special part of the `CREATE INDEX` statement.

I think the best way would be to exclude `Collate` from `IndexExpression.wrapper_classes` on MySQL, that would cause the `COLLATE` part to be wrapped with parentheses. Not sure if we can use `IndexExpression.register_wrappers` for the exclusion though..

Reply:
Unfortunately, we cannot remove `Collate` from `wrapper_classes` because it will break, e.g.
```python
Collate(F('title').desc(), collation=collation)
```

Reply:
Right, wrappers need to be top-most expressions. I agree that this will would be a pretty weird gotcha on MySQL where the `desc()` would need to be called on `Collate` instead. But maybe that's OK, we do provide an error message explaining this. We could pass `connection` to `IndexExpression` and filter `wrapper_classes` depending on support of the `COLLATE` clause for `CREATE INDEX`.

Reply:
OK I pushed fix.